pkg/restapi/vc/operation/operations.go

```diff
@@ -322,41 +361,90 @@ func (o *Operation) storeVCHandler(rw http.ResponseWriter, req *http.Request) {
 		return
 	}
 
-	doc := operation.StructuredDocument{}
-	doc.Content = make(map[string]interface{})
-	doc.Content["message"] = data.Credential
+	o.storeVC(data, vc, rw)
+}
 
-	// We need an EDV-compliant ID. The UUID that's in the VC is a good fit, since it's 128 bits long already.
-	// Let's convert it to base58 and use it as the StructuredDocument ID.
-	edvDocID, err := convertVCIDToEDVCompatibleFormat(vc.ID)
+func (o *Operation) storeVC(data *StoreVCRequest, vc *verifiable.Credential, rw http.ResponseWriter) {
+	doc, err := o.buildStructuredDoc(data, vc)
 	if err != nil {
 		o.writeErrorResponse(rw, http.StatusBadRequest, err.Error())
 
 		return
 	}
 
-	doc.ID = edvDocID
+	marshalledStructuredDoc, err := json.Marshal(doc)
+	if err != nil {
+		o.writeErrorResponse(rw, http.StatusBadRequest, err.Error())
+
+		return
+	}
+
+	// We have no recipients, so we pass in the sender key as the recipient key as well
+	encryptedStructuredDoc, err := o.packer.Pack(marshalledStructuredDoc,
+		base58.Decode(o.senderKey), [][]byte{base58.Decode(o.senderKey)})
+	if err != nil {
+		o.writeErrorResponse(rw, http.StatusBadRequest, err.Error())
+
+		return
+	}
+
+	encryptedDocument := operation.EncryptedDocument{
+		ID:       doc.ID,
+		Sequence: 0,
+		JWE:      encryptedStructuredDoc,
+	}
+
+	_, err = o.client.CreateDocument(data.Profile, &encryptedDocument)
+	if err != nil {
+		o.writeErrorResponse(rw, http.StatusBadRequest, err.Error())
+
+		return
+	}
 
-	_, err = o.client.CreateDocument(data.Profile, &doc)
+	err = o.idMapping.Put(vc.ID, []byte(doc.ID))
 	if err != nil {
 		o.writeErrorResponse(rw, http.StatusBadRequest, err.Error())
 
 		return
 	}
 }
 
-// This function expects the vcID that's passed in to be a URL with a "/" and then a UUID concatenated to the end.
-func convertVCIDToEDVCompatibleFormat(vcID string) (string, error) {
-	vcIDParts := strings.Split(vcID, "/")
+func (o *Operation) buildStructuredDoc(data *StoreVCRequest,
+	vc *verifiable.Credential) (*operation.StructuredDocument, error) {
+	var edvDocID string
 
-	uuidFromVCID := vcIDParts[len(vcIDParts)-1]
+	idFromMapping, err := o.idMapping.Get(vc.ID)
+	if err != nil {
+		edvDocID, err = generateEDVCompatibleID()
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		edvDocID = string(idFromMapping)
+	}
 
-	parsedUUID, err := uuid.Parse(uuidFromVCID)
+	doc := operation.StructuredDocument{}
+	doc.ID = edvDocID
```

Comment:
Shouldn't there be a check here `if edvDocID == ""` in case err was nil but edvDocID returned was empty above ? 

I guess this is another reason this logic needs to be revisited.

Reply:
I  don't believe it's possible for edvDocID to be blank at this point, unless someone manually put a blank string in the database. But even if it was blank, the EDV would just return an error, which would cause the store VC handler to return an error back to the person who called that endpoint.